### ISSUE TITLE
[IDEA] Provide a way to send code block to buffers

### ISSUE THREAD

**Issue by m-lce** [chep/copilot-chat.el#59]
First of all thanks for the plugin!
I wonder if you are thinking about a way to provide a way to send the code generated in chat to buffers to add or replace existing code. Something showing the diff would be amazing.
I think i saw something like this in the neovim plugin that inspired this one.

**Comment by chep** [comment]
Hi,
Have you tried `(copilot-chat-ask-and-insert)`?

**Comment by m-lce** [comment]
Yes but the issue with that it’s the llm response is not code only.
Or maybe i missed something ...

**Comment by kassick** [comment]
IMHO that's something nice to have in the chat response buffer.

One may have asked something, clarified, asked for improvements, etc. That's not what ask-and-insert provides.

Considering that the code portions of the response should be code-fenced (either with triple-backquote in markdown or with begin-src in org), we could bind some key in the buffer to find code inside the org or markdown element and either copy or insert it in a buffer.

But what would be the target buffer to insert? If I have the chat windows already visible and I switch to them with the mouse, should I insert in the last active window? What if I call `other-window` repeatedly to get to the chat window?

**Comment by kassick** [comment]
(that said, since I use evil-mode, I frequently use some `yank inner` something to copy the generated code and paste it in the right place

**Comment by chep** [comment]
Here are some useful functions that can be used. I do not have time to integrate them for the moment but I need to save them somewhere :

org:
```
(defun get-org-block-content-at-point ()
  "Get the content of the org block at point."
  (let* ((element (org-element-at-point))
         (type (org-element-type element)))
	(print type t)
    (when (memq type '(src-block quote-block example-block))
      (let ((content (org-element-property :value element)))
        content))))
```

markdown (generated by copilot, not tested)
````
(defun get-markdown-block-content-at-point ()
  "Get the content of the markdown code block at point."
  (when (markdown-code-block-at-point-p)
    (let* ((bounds (markdown-get-enclosing-fenced-block-construct))
           ;; bounds is (begin-fence . end-fence)
           (begin-fence (car bounds))
           (end-fence (cdr bounds))
           ;; Get positions after/before fences
           (content-start (save-excursion 
                           (goto-char begin-fence)
                           (forward-line)
                           (point)))
           (content-end (save-excursion
                         (goto-char end-fence)
                         (forward-line 0)
                         (point))))
      (when (and content-start content-end)
        ;; Get language from the opening fence
        (let* ((lang (save-excursion
                      (goto-char begin-fence)
                      (markdown-code-block-lang)))
               (content (buffer-substring-no-properties 
                        content-start content-end)))
          (list :content content
                :language lang))))))
```

**Comment by chep** [comment]
I pushed a new commit with the `copilot-chat-send-to-buffer` function. It sends code block at point to a buffer. If there is a selected region in the destination buffer, it is replaced by the code.

**Comment by chep** [comment]
By the way it only works with org-mode for the moment.

**Comment by m-lce** [comment]
Thank you very much, it works well. 
The next level would be to detect automatically the buffer and where to insert the code block but that’s great to already have this !

**Comment by chep** [comment]
The last visited buffer using major mode corresponding to the code blocks is now the first one in completion list.

(The basic version also works for markdown)

**Comment by m-lce** [comment]
Thanks a lot for the update. 
After using it a little, i think a copilot-chat-insert-block called from the buffer would be cool. 
The idea is when you call this function you can choose among the list of code blocks of the chat buffer.
Tell me what you think about that.

**Comment by chep** [comment]
That could be great. Maybe I could do something "kill ring like", with next and previous functions.

**Comment by chep** [comment]
If you are using org frontend, you now have `copilot-chat-yank` and `copilot-chat-yank-pop` functions available. They act like `yank` and `yank-pop`.
I use them like this :
``` emacs-lisp
(use-package copilot-chat
  …
  :bind
  ("C-c C-y" . copilot-chat-yank)
  ("C-c M-y" . copilot-chat-yank-pop)
  ("C-c C-M-y" . (lambda () (interactive) (copilot-chat-yank-pop -1))))
```

**Comment by m-lce** [comment]
Thanks, i will give it a try !
